llama-index-core/tests/workflow/test_service.py

```diff
@@ -0,0 +1,34 @@
+import pytest
+
+from llama_index.core.workflow.decorators import step
+from llama_index.core.workflow.events import Event, StartEvent, StopEvent
+from llama_index.core.workflow.workflow import Workflow
+
+
+class ServiceWorkflow(Workflow):
+    @step()
+    async def generate(self, ev: StartEvent) -> StopEvent:
+        return StopEvent(result=42)
+
+
+class NumGenerated(Event):
+    num: int
+
+
+class DummyWorkflow(Workflow):
+    @step(services=["service_workflow"])
+    async def get_a_number(self, svc: Workflow, ev: StartEvent) -> NumGenerated:
+        res = await svc.run()
+        return NumGenerated(num=int(res))
+
+    @step()
+    async def multiply(self, ev: NumGenerated) -> StopEvent:
+        return StopEvent(ev.num * 2)
+
+
+@pytest.mark.asyncio()
+async def test_e2e():
+    wf = DummyWorkflow()
+    wf.add_service("service_workflow", ServiceWorkflow())
```

Comment:
It feels somewhat unintuituve that this isn't a constructor argument 🤔 (I think this is similar to how the steps worked right?)

Reply:
This is a relic of a first attempt when I tried to attach the services to the workflow class, same as the steps. We can definitely add it to the constructor